docs/exceptions.md

```diff
@@ -164,13 +195,29 @@ to return `Maybe v`. In 100% pure code, use one of these ways to handle errors:
 * `MaybeT`, `ExceptT e`
 * `CatchT`
 
+*Comment:* neither of these examples are of errors or exceptions! They are
+examples of normal values in the images of pure functions. The `Map` does not
+make any guarantee to that any particular key is in the map! Nor does the parser
+guarantee that every input will be recognized! These facts are reflected in the
+types. This has *nothing at all* to do with errors or exceptions. We need to
+strike this section out.
```

Comment:
As the primary author of the document I want to make something clear: `lookup :: k -> Map k v -> Maybe v` is the *right* type for this operation, in full compliance with the guidelines. What does it have to do with Python exceptions?

Reply:
What about `Database.lookup :: DbConn -> k -> IO ???`? I would think that `??? ~ Maybe a`. Even though you *can* throw an exception in `IO` for the missing record, it seems like less of an exceptional situation that eg the database connection being refused, or the network connection screwing up. 

The type `IO (Either DatabaseError (Maybe a))` would be cumbersome, because we're generally just going to want to rethrow the `DatabaseError` when we're dealing with code that wants to handle `a`.

I think the way I have approached this in  the past is to consider: "if this operation could be done purely, in some absurd way, how would I type it?" For a database, we can assume we have some `Datatabase` type, that is the entire database that we're passing around. The resulting type is `lookup :: Database -> k -> Maybe a`. When we add impurity, we get `lookup :: DatabaseConn -> k -> IO (Maybe a)`. 

If I were to reimagine `readFile`, I'd say: "Imagine we have the entire file system in memory", giving us `readFile :: FileSystem -> FileName -> Maybe FileContents`. Impurifying it, we get `readFile :: FileName -> IO (Maybe FileContents)`.  This handles the exceedingly common "file not found" error, while hiding details like "no more RAM available", "no more open file handles available," etc.

Reply:
> As the primary author of the document I want to make something clear: lookup :: k -> Map k v -> Maybe v is the right type for this operation, in full compliance with the guidelines. We use Maybe to represent this unwanted codepath, classified as "regular error" by the guidelines. What does it have to do with Python exceptions?

My point is that this has nothing at all to do with exceptions and so shouldn't even be mentioned in the document. There's no "unwanted codepath" here.

Reply:
@parsonsmatt I'm fine with having `IO (Either DatabaseError (Maybe a))` as long as `DatabaseError` doesn't have an `Exception` instance. If it has an `Exception` instance, it means that it might be both thrown as `IO` exception and returned in `Either`, and it's a headache at each call site. I'm also opposed to `ExceptT DatabaseError IO a` because it means we have two exit paths from the monad (`throwM` and `throwError`) and there's no formal criteria which one to use for what.

Reply:
That makes sense to me. Thanks for clarifying!

Reply:
@avieth Exceptions *are* unwanted codepaths. "Exception" can be read as "exceptional situation". You're using a different definition, apparently, but I already asked you to treat the document on its own terms, because otherwise this discussion is unproductive.

Reply:
Not *unwanted*, just *exceptional*. Exceptions are not "code paths", they're terms in the object language. Further, a parse error, or a map lookup in which the key is not present, are neither unwanted nor exceptional. 

> I already asked you to treat the document on its own terms, because otherwise this discussion is unproductive.

I have objections to the document itself. It's not immune to criticism. And if it's too be taken as some sort of law that can be cited in other pull request discussions, then it's very important that it be criticised.

Reply:
I am confused, are your objections to the meaning, the form, or both? Let's tackle these separately, because otherwise the discussion of meaning drowns in a sea of nitpicks to form. So if you have better phrasing in mind, that's one thing, and if you want the sentence to say something completely different, it's another. What is it, then?

Reply:
Objections to meaning and to what it prescribes. Some definitions are vague and ultimately unhelpful (unwanted, secondary) and the prescriptions favour throwing exceptions too liberally.